contracts/src/v0.8/dev/VRFCoordinatorV2.sol

```diff
@@ -0,0 +1,541 @@
+    // are, otherwise they could use someone else's subscription balance.
+    if (s_subscriptions[subId].consumers[consumerId] != msg.sender) {
+      revert InvalidConsumer(msg.sender);
+    }
+    // Input validation using the config storage word.
+    if (minimumRequestConfirmations < s_config.minimumRequestBlockConfirmations) {
+      revert RequestBlockConfsTooLow(minimumRequestConfirmations, s_config.minimumRequestBlockConfirmations);
+    }
+    if (s_subscriptions[subId].balance < s_config.minimumSubscriptionBalance) {
+      revert InsufficientBalance();
+    }
+    if (callbackGasLimit > s_config.maxGasLimit) {
+      revert GasLimitTooBig(callbackGasLimit, s_config.maxGasLimit);
+    }
+    // We could additionally check s_serviceAgreements[keyHash] != address(0)
+    // but that would require reading another word of storage. To save gas
+    // we leave that out.
+    uint256 nonce = s_nonces[keyHash][msg.sender] + 1;
+    uint256 preSeedAndRequestId = uint256(keccak256(abi.encode(keyHash, msg.sender, nonce)));
+
+    s_callbacks[preSeedAndRequestId] = keccak256(abi.encodePacked(
+        preSeedAndRequestId,
+        block.number,
+        subId,
+        callbackGasLimit,
+        numWords,
+        msg.sender));
+    emit RandomWordsRequested(keyHash, preSeedAndRequestId, subId, minimumRequestConfirmations, callbackGasLimit, numWords, msg.sender);
+    s_nonces[keyHash][msg.sender] = nonce;
+
+    return preSeedAndRequestId;
+  }
+
+  function getCallback(
+      uint256 requestId
+  )
+    external
+    view
+    returns (
+      bytes32
+    )
+  {
+    return s_callbacks[requestId];
+  }
+
+  // Offsets into fulfillRandomnessRequest's proof of various values
+  //
+  // Public key. Skips byte array's length prefix.
+  uint256 public constant PUBLIC_KEY_OFFSET = 0x20;
+  // Seed is 7th word in proof, plus word for length, (6+1)*0x20=0xe0
+  uint256 public constant PRESEED_OFFSET = 0xe0;
+
+  function fulfillRandomWords(
+    bytes memory proof
+  )
+    external
+  {
+    uint256 startGas = gasleft();
+    (bytes32 keyHash, uint256 requestId,
+    uint256 randomness, FulfillmentParams memory fp) = getRandomnessFromProof(proof);
+
+
+    uint256[] memory randomWords = new uint256[](fp.numWords);
+    for (uint256 i = 0; i < fp.numWords; i++) {
+      randomWords[i] = uint256(keccak256(abi.encode(randomness, i)));
+    }
+
+    // Prevent re-entrancy. The user callback cannot call fulfillRandomWords again
+    // with the same proof because this getRandomnessFromProof will revert because the requestId
+    // is gone.
+    delete s_callbacks[requestId];
+    VRFConsumerV2Interface v;
+    bytes memory resp = abi.encodeWithSelector(v.fulfillRandomWords.selector, requestId, randomWords);
+    uint256 gasPreCallback = gasleft();
+    if (gasPreCallback < fp.callbackGasLimit) {
+      revert InsufficientGasForConsumer(gasPreCallback, fp.callbackGasLimit);
+    }
+    (bool success,) = fp.sender.call(resp);
```

Comment:
It appears ```call{gas: fp.callbackGasLimit}(resp)``` does the same thing as ```callWithExactGas``` so I've added that

Reply:
`call{gas: fp.callbackGasLimit}(resp)` will ensure that the callee has _at most_ `fp.callbackGasLimit` gas available. So it does indeed prevent the callee from using too much gas. However, it does not guarantee that the callee actually gets `fp.callbackGasLimit`, they could get less. `callWithExactGas` should give the callee exactly the amount of gas specified.

This contract currently does have some checks that try to ensure that the callee gets at least `fp.callbackGasLimit`:
```
    if (gasPreCallback < fp.callbackGasLimit) {
      revert InsufficientGasForConsumer(gasPreCallback, fp.callbackGasLimit);
    }
```
However, these checks are less precise than `callWithExactGas`. They do not take into account the 63/64 relative cap and they also do not consider the cost of the call to `fp.sender` itself. So the code as currently written could well lead to the requester receiving a couple of thousand gas less than they expect.

TL;DR: `callWithExactGas` is a better fit here

Reply:
Oh hm I noticed that even with callWithExactGas they still get less than they expect inside their callback because of additional byte code involved in actually making it to the users code (aside from just the call op code itself and relative op code). In particular, what they end up with inside their callback is dependent on the arguments, which in the case of the fulfillRandomWords could vary quite widely since randomWords is a dynamic array whose size is up to the user. Seemed cleaner to not attempt to ensure a call to gasleft inside their callback is exactly what they requested

Reply:
Not sure I'm understanding you correctly, but the byte code you mention is on the callee side, no? (Compiler generated argument decoding, etc...)

Reply:
yeah the byte code is on the callee side, sorry by "user code" I meant like the solidity that they would write inside fulfillRandomWords

Reply:
Okay, sounds like we're on the same page.

> Seemed cleaner to not attempt to ensure a call to gasleft inside their callback is exactly what they requested and since in v2 they control the callback gas budget, let them do the buffering if they really need it

Agreed!

Reply:
To be clear, I _do_ think that we should use callWithExactGas. Their callframe gets the exact amount of gas requested and then what they do inside it (e.g. argument decoding) is their own business.

Reply:
Fair enough. I think if a user is sophisticated enough to actually account for the gas costs of argument decoding, they could probably account for the call + 63/64, but we may as well make that buffer calculation easier with callWithExactGas - I'll add it.